torch/csrc/inductor/aoti_torch/shim_common.cpp

```diff
@@ -0,0 +1,528 @@
+    AOTITensorManagerHandle tensor_manager) {
+  CONVERT_EXCEPTION_TO_ERROR_CODE({
+    TensorManager* manager = reinterpret_cast<TensorManager*>(tensor_manager);
+    delete manager;
+  });
+}
+
+AOTITorchError aoti_torch_extern_tensor_to_aoti_tensor(
+    AOTITensorManagerHandle tensor_manager,
+    AOTITensorHandle* aoti_tensor,
+    AtenTensorHandle aten_tensor) {
+  CONVERT_EXCEPTION_TO_ERROR_CODE({
+    at::Tensor* atensor = reinterpret_cast<at::Tensor*>(aten_tensor);
+    *aoti_tensor = create_managed_atoi_tensor(
+        tensor_manager,
+        atensor->data_ptr(),
+        // Input and output tensors are managed by the external runtime, so we
+        // set the deleter as at::detail::noopDelete to prevent the tensor
+        // storage from being freed, and also the size and stride pointers are
+        // guaranteed to be valid.
+        (void*)at::detail::noopDelete,
+        atensor->dim(),
+        atensor->sizes().data(),
+        atensor->strides().data(),
+        0, // atensor->data_ptr() has added storage_offset, so set to 0 here
+        static_cast<int32_t>(atensor->device().type()),
+        static_cast<int32_t>(atensor->device().index()),
+        static_cast<int32_t>(atensor->scalar_type()));
+  });
+}
+
+AOTITorchError aoti_torch_free_tensor_storage(AOTITensorHandle aoti_tensor) {
+  CONVERT_EXCEPTION_TO_ERROR_CODE({
+    AOTITensor* t = reinterpret_cast<AOTITensor*>(aoti_tensor);
+    TORCH_CHECK(
+        t != nullptr && t->data_ptr && t->deleter != nullptr,
+        "aoti_torch_free_tensor_storage: Invalid AOTITensorHandle")
+    c10::DeleterFnPtr deleter = reinterpret_cast<c10::DeleterFnPtr>(t->deleter);
+    deleter(t->data_ptr);
+    t->data_ptr = nullptr; // avoid double free
+  });
+}
+
+AOTITorchError aoti_torch_get_data_ptr(
+    AOTITensorHandle aoti_tensor,
+    void** data_ptr) {
+  CONVERT_EXCEPTION_TO_ERROR_CODE({
+    AOTITensor* t = reinterpret_cast<AOTITensor*>(aoti_tensor);
+    char* base = static_cast<char*>(t->data_ptr);
+    // t->storage_offset is in the number of elements, so turn it into bytes
+    size_t offset_in_bytes = t->storage_offset * get_dtype_bytes(t->dtype);
+    *data_ptr = (void*)(base + offset_in_bytes);
+  });
+}
+
+AOTITensorHandle aoti_torch__reinterpret_tensor(
+    AOTITensorManagerHandle tensor_manager,
+    AOTITensorHandle self,
+    int64_t ndim,
+    const int64_t* sizes_ptr,
+    const int64_t* strides_ptr,
+    int64_t offset_increment) {
+  AOTITensor* t = reinterpret_cast<AOTITensor*>(self);
+  return create_managed_atoi_tensor(
+      tensor_manager,
+      t->data_ptr,
+      t->deleter,
+      ndim,
+      sizes_ptr,
+      strides_ptr,
+      t->storage_offset + offset_increment,
+      t->device_type,
+      t->device_index,
+      t->dtype);
+}
+
+// TODO: implement a more efficient version instead of calling into aten
+AOTITorchError aoti_torch_empty_strided(
+    AOTITensorManagerHandle tensor_manager,
+    AOTITensorHandle* out,
```

Comment:
It's weird that the output argument is not at the front or end (which would be conventional in C)